wagtail/wagtailcore/models.py

```diff
@@ -1277,17 +1283,184 @@ def serve_password_required_response(self, request, form, action_url):
+        else:
+            # If the User has no explorable pages, display nothing.
+            return Page.objects.none()
+
+    def get_explorable_ancestors(self, user):
+        """
+        Returns a queryset of all the node's ancestors that the given user is permitted to see in the Explorer.
+        """
+        # Superusers see all the ancestors.
+        if user.is_superuser:
+            return self.get_ancestors()
+
+        # Copied this algorithm from TreeQuerySet.ancestor_of_q, but tweaked it to remove this page's path.
+        ancestor_paths = [
+            self.path[0:pos]
+            for pos in range(0, len(self.path) + 1 - self.steplen, self.steplen)[1:]
+        ]
+        # Remove all paths that are neither permitted nor required.
+        permitted_paths, required_ancestors = get_explorable_page_paths(user)
+        for path in ancestor_paths[:]:
+            if path not in permitted_paths and path not in required_ancestors:
+                ancestor_paths.remove(path)
+
+        return Page.objects.filter(path__in=ancestor_paths)
+
+    def is_explorable_root(self, user):
+        # A superuser's explorable root is the Root page.
+        if user.is_superuser:
+            return self.is_root()
+
+        permitted_paths, required_ancestors = get_explorable_page_paths(user)
+        return self.path == required_ancestors[0]
+
     class Meta:
         verbose_name = _('page')
         verbose_name_plural = _('pages')
 
 
-def get_navigation_menu_items():
+def get_explorable_page_paths(user):
+    """
+    Returns a tuple of lists of treebeard "path" values.
+    The first list is the set of paths which the given user is explicitly permitted to administer (meaning they can
+    administer all children of those paths, too).
+    The second list is the paths which must be shown to the user in order for them to navigate to their permitted Pages
+    in the Explorer. Unlike the paths in the first list, the children of these paths are not to be automatically made
+    visible.
+
+    NOTE: The first element of the second list is always the Closest Common Ancestor of the User's permitted Pages.
+    """
+    # We cache the output of this function because it's very expensive to generate it fresh every time the user loads
+    # an Explorer page. Unfortunately, this does lead to a lot of duplciated data in the cache, as every single user
+    # that has the same set of groups will have the same permitted paths. I'd like to cache by groupset, which would
+    # remove all that duplication, but that would require an extra DB call every time, to get the user's group list.
+    cache_key = EXPLORABLE_PATHS_PREFIX.format(user.pk)
+    permitted_paths, required_ancestors = cache.get(cache_key, ([], []))
+    if not permitted_paths:
+        ###################
+        # Get paths by User
+        ###################
+        permissions = (
+            GroupPagePermission.objects
```

Comment:
Can we put all logic related to permission into `UserPagePermissionsProxy` and `PagePermissionTester`?